### PR TITLE
Add IndexBackfiller to Support Client Side Indexing

### PR DESCRIPTION
**Pull request by cherylEnkidu**
#no-changelog

### CODE REVIEW

**Review by Eldeloro1 — APPROVED** [review]

File: Firestore/core/src/local/leveldb_remote_document_cache.cc [hunk 1]
```diff
@@ -170,115 +171,82 @@ MutableDocumentMap LevelDbRemoteDocumentCache::GetAll(
...
-    DocumentKeySet remote_keys;
-
-    LevelDbRemoteDocumentReadTimeKey current_key;
-    for (; it->Valid() && current_key.Decode(it->key()); it->Next()) {
-      const ResourcePath& collection_path = current_key.collection_path();
-      if (collection_path != path) {
+  MutableDocumentMap result;
+  for (auto path = collections.cbegin();
+       path != collections.cend() && result.size() < limit; path++) {
+    const auto map = GetAll(*path, offset);
```
Comment (wu-hui) [inline Firestore/core/src/local/leveldb_remote_document_cache.cc:200]: please give it a better name than map

File: Firestore/core/src/local/leveldb_remote_document_cache.cc [hunk 2]
```diff
@@ -170,115 +171,82 @@ MutableDocumentMap LevelDbRemoteDocumentCache::GetAll(
...
 MutableDocumentMap LevelDbRemoteDocumentCache::GetAllExisting(
-    const DocumentKeySet& keys) {
-  MutableDocumentMap docs = LevelDbRemoteDocumentCache::GetAll(keys);
+    const DocumentVersionMap key_version_map) const {
   MutableDocumentMap result;
-  for (const auto& kv : docs) {
-    const DocumentKey& key = kv.first;
-    auto& document = kv.second;
+  for (const auto& pair : key_version_map) {
+    auto document = Get(pair.first).WithReadTime(pair.second);
```
Comment (wu-hui) [inline Firestore/core/src/local/leveldb_remote_document_cache.cc:177]: We probably should use the background queue to do the deserialization: see line 144 for an example.

File: Firestore/core/src/local/leveldb_remote_document_cache.cc [hunk 3]
```diff
@@ -170,115 +171,82 @@ MutableDocumentMap LevelDbRemoteDocumentCache::GetAll(
...
-    for (; it->Valid() && current_key.Decode(it->key()); it->Next()) {
-      const ResourcePath& collection_path = current_key.collection_path();
-      if (collection_path != path) {
+  MutableDocumentMap result;
+  for (auto path = collections.cbegin();
+       path != collections.cend() && result.size() < limit; path++) {
+    const auto map = GetAll(*path, offset);
+    for (const auto& pair : map) {
+      result = result.insert(pair.first, pair.second);
+      if (result.size() == limit) {
```
Comment (wu-hui) [inline Firestore/core/src/local/leveldb_remote_document_cache.cc:203]: Unlike other platforms (which iterates over resultset), here you call `GetAll` for all documents, then throw away some if the result is over the limit.

You can pass `limit` down to `GetAll(*path, offset)` and make sure we only process enough documents.

**Review by wu-hui — COMMENTED** [review]
I have not reviewed the tests yet, want to give you a first round of feedback.

File: Firestore/core/test/unit/local/index_backfiller_test.cc [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+  }
+
+  void VerifyQueryResults(
+      const std::string& collection_group,
+      const std::unordered_set<std::string>& expected_keys) const {
+    VerifyQueryResults(Query(collection_group).AddingOrderBy(OrderBy("foo")),
+                       expected_keys);
+  }
+
+  /** Adds a set mutation to a batch with the specified id for every specified
```
Comment (wu-hui) [inline Firestore/core/test/unit/local/index_backfiller_test.cc:140]: Format

File: Firestore/core/test/unit/local/index_backfiller_test.cc [hunk 5]
```diff
@@ -0,0 +1,559 @@
...
+  RemoteDocumentCache* remote_document_cache_;
+  DocumentOverlayCache* document_overlay_cache_;
+  CountingQueryEngine query_engine_;
+  LocalStore local_store_;
+  IndexManager* index_manager_;
+  IndexBackfiller* index_backfiller_;
+};
+
+TEST_F(IndexBackfillerTest,
+       BackfillWritesLatestReadTimeToFieldIndexOnCompletion) {
```
Comment (wu-hui) [inline Firestore/core/test/unit/local/index_backfiller_test.cc:173]: Simply `WritesLatestReadTimeToFieldIndexOnCompletion`, applies to all below.

**Review by wu-hui — APPROVED** [review]
Approved with some nits, great work!!

**Pull request merged**